File: audiobook/src/main/java/de/ph1b/audiobook/utils/FileRecognition.java
package de.ph1b.audiobook.utils;

import android.os.Build;

import java.io.FileFilter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


/**
 * Class containing methods for recognizing different file types by their file ending.
 *
 * @author Paul Woitaschek
 */
public class FileRecognition {

    private static final List<String> AUDIO_TYPES = new ArrayList<>(30);
    public static final FileFilter FOLDER_AND_MUSIC_FILTER = pathname -> {
        for (String s : AUDIO_TYPES) {
            if (pathname.getName().toLowerCase().endsWith(s)) {
                return true;
            }
        }
        return pathname.isDirectory();
    };
    private static final List<String> IMAGE_TYPES = Arrays.asList(".jpg", ".jpeg", ".png", ".bmp");
    public static final FileFilter FOLDER_AND_IMAGES_FILTER = pathname -> {
        for (String s : IMAGE_TYPES) {
            if (pathname.getAbsolutePath().toLowerCase().endsWith(s)) {
                return true;
            }
        }
        return pathname.isDirectory();
    };

    static {
        AUDIO_TYPES.add(".3gp");

        AUDIO_TYPES.add(".aac");
        AUDIO_TYPES.add(".awb");

        AUDIO_TYPES.add(".flac");

        AUDIO_TYPES.add(".imy");

        AUDIO_TYPES.add(".m4a");
        AUDIO_TYPES.add(".m4b");
        AUDIO_TYPES.add(".mp4");
        AUDIO_TYPES.add(".mid");
        AUDIO_TYPES.add(".mkv");
        AUDIO_TYPES.add(".mp3");
        AUDIO_TYPES.add(".mp3package");
        AUDIO_TYPES.add(".mxmf");

        AUDIO_TYPES.add(".ogg");
        AUDIO_TYPES.add(".oga");
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            AUDIO_TYPES.add(".opus");
        }
        AUDIO_TYPES.add(".ota");

        AUDIO_TYPES.add(".rtttl");
        AUDIO_TYPES.add(".rtx");

        AUDIO_TYPES.add(".wav");
        AUDIO_TYPES.add(".wma");

        AUDIO_TYPES.add(".xmf");
    }
}

File: audiobook/src/main/kotlin/de/ph1b/audiobook/utils/FileRecognition.kt
/*
 * This file is part of Material Audiobook Player.
 *
 * Material Audiobook Player is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or any later version.
 *
 * Material Audiobook Player is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Material Audiobook Player. If not, see <http://www.gnu.org/licenses/>.
 * /licenses/>.
 */

package de.ph1b.audiobook.utils

import android.os.Build
import com.google.common.collect.Lists
import com.google.common.io.Files
import java.io.FileFilter
import java.util.*


/**
 * Class containing methods for recognizing different file types by their file ending.

 * @author Paul Woitaschek
 */
object FileRecognition {

    private val imageTypes = Arrays.asList("jpg", "jpeg", "png", "bmp")
    private val audioTypes: List<String>

    init {
        audioTypes = Lists.newArrayList("3gp",
                "aac",
                "awb",
                "flac",
                "imy",
                "m4a",
                "m4b",
                "mp4",
                "mid",
                "mkv",
                "mp3",
                "mp3package",
                "mxmf",
                "ogg",
                "oga",
                "ota",
                "rtttl",
                "rtx",
                "wav",
                "wma",
                "xmf"
        )
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            audioTypes.add("opus")
        }
    }

    val folderAndMusicFilter = FileFilter {
        if (it.isDirectory) {
            return@FileFilter true
        } else {
            val extension = Files.getFileExtension(it.name)
                    .toLowerCase()
            return@FileFilter audioTypes.contains(extension)
        }
    }
    val folderAndImagesFilter = FileFilter {
        if (it.isDirectory) {
            return@FileFilter true
        } else {
            val extension = Files.getFileExtension(it.name)
                    .toLowerCase()
            return@FileFilter imageTypes.contains(extension)
        }
    }
}